### ISSUE TITLE
[Question] - use incoming connection credentials for proxy

### ISSUE THREAD

**Issue by haganbt** [sidorares/node-mysql2#1163]
When using `mysql2` as a [proxy](https://github.com/sidorares/node-mysql2/blob/master/examples/mysqlproxy.js), is there anyway to use the incoming `password` credential in the downstream destination connection?

For example, in the below extract we have the incoming connection `conn`. I would like to use the username and password from that within the `remote` connection:

```javascript
server.on('connection', conn => {
  console.log('connection');

  conn.serverHandshake({
    protocolVersion: 10,
    serverVersion: 'node.js rocks',
    connectionId: 1234,
    statusFlags: 2,
    characterSet: 8,
    capabilityFlags: 0xffffff ^ ClientFlags.COMPRESS
  });

  conn.on('field_list', (table, fields) => {
    console.log('field list:', table, fields);
    conn.writeEof();
  });

  const remote = mysql.createConnection({
    user: 'root',
    database: 'dbname',
    host: 'server.example.com',
    password: 'secret'
  });
...
```

I see that within the `conn.serverHandshake` I can use `authCallback` which does give me the `user` as well as a few other properties like `authToken`:

```javascript
{
  user: 'my_user',
  database: 'my_database',
  address: '::ffff:127.0.0.1',
  authPluginData1: <Buffer 94 2b 93 30 a2 88 08 4c>,
  authPluginData2: <Buffer 3e 77 00 c5 00 1a f2 86 96 ad 21 30>,
  authToken: <Buffer 50 3c 77 a6 2c cf e4 15 e5 7a 7f 41 9f 0e 53 de a7 21 6d bf>
}
```

Can the password be extracted in any way to use as part of the downstream destination connection?

Appreciate any guidance.

**Comment by sidorares** [comment]
client does not send cleartext password (at least with `mysql_native_password` auth), you can only have access to `sha1(password)`

**Comment by sidorares** [comment]
tough you can still pass sha1 to downstream, that should be enough to authenticate - https://github.com/sidorares/node-mysql2/blob/0c913ea965ed5c1eb26ed37064bce6f1e38d2095/examples/pass-sha.js#L7

**Comment by haganbt** [comment]
Great stuff, thanks @sidorares. Just to make sure I fully understand, that means I can use the `authToken` property and send that downstream using `passwordSha1`?

**Comment by sidorares** [comment]
yes, (maybe not that directly). Take a look at how token is calculated from password and salt - https://github.com/sidorares/node-mysql2/blob/c955041b0785bb7b7cc5536982fab0a418e70a88/lib/auth_41.js#L60-L81

**Comment by haganbt** [comment]
Got it, super helpful. Thank you!

**Comment by MartinGassner** [comment]
> you can only have access to `sha1(password)`

I was wondering how I can access the `sha1(password)` within the serverHandshake to authenticate for a new connection. @sidorares maybe you can help?

Is it possible to get the `sha1(password)` with the "authToken, authPluginData1 and authPluginData2"?

**Comment by sidorares** [comment]
> Is it possible to get the sha1(password) with the "authToken, authPluginData1 and authPluginData2"?

probably not possible, looks very one way

salt = authPluginData1 + authPluginData2
token = sha1(salt xor sha1(sha1(password)))

You can generate token if you know sha1(password) or sha1(sha1(password))

**Comment by MartinGassner** [comment]
Ok then I assume there is no way to access the `password` or `passwordSha1` within the ServerHandshake?

**Comment by sidorares** [comment]
@haganbt @MartinGassner it's getting a bit confusing with 3 parities ( client, proxy server, "real" server ). Can you describe in a bit more details what and where you want to validate?

|sequence No | Client   |      Proxy      |  Server |
|-|----------|:-------------:|------:|
| 0 |  | <- salt  |  <- salt |
| 1 | `username, token = sha1(salt xor sha1(sha1(password)))` -> |    |   |

If proxy stores `username, sha1(password)` it can validate the client and also pass down token unchanged so that "real" server can also accept connection for given salt,token combination

**Comment by MartinGassner** [comment]
we trying to create a mysql proxy without storing `user` or `password` so if a new connection comes in the `user` and `password` from that connection are passed through to the "real" connection. The proxy wouldn't need to store any user information and therefor we wouldn't need to validate anything.

So far the theory, but it seems we have no access to the `password` or `passwordSha1` within the ServerHandshake. So I think we'll need to store the `user` and at least the `passwordSha1` somewhere so we can validate an incoming connection to the proxy with the  `authToken`, `authPluginData1` and `authPluginData2`

**Comment by sidorares** [comment]
yes, if you want to validate incoming connections before you allow them to connect to the server you'll need to access sha1(password) from the proxy ( or you can delegate validation to real server if it's 1:1 auth mapping between proxy and server )

**Comment by MartinGassner** [comment]
ok thank you very much @sidorares !

**Comment by haganbt** [comment]
Hi @sidorares apologies for re-opening this issue. As I re-read, I just want to make sure I am really clear on our options and fully understand.

The goal is for an incoming client connection to pass through the proxy and authenticate with the 'real' destination db. Lets assume we dont validate at the proxy and we dont store the password at the proxy. 

>that means I can use the authToken property and send that downstream using passwordSha1?

>yes, (maybe not that directly). Take a look at how token is calculated from password and salt -

From this comment, it implies this is possible. Is that accurate or were we assuming validation etc at the proxy?

The goal is to not store the password in any way at the proxy, just pass it through to the destination.

Really appreciate any comments, this is a real blocker for our product :)

**Comment by sidorares** [comment]
> The goal is to not store the password in any way at the proxy, just pass it through to the destination.

it's not possible to extract password ( or validate on the proxy if you don't know password or sha1(password) or sha1(sha1(password)), but if you just transparently send same salt to the client ( basically copy server hello packet as is ) and return same response token to real db you should be fine.

Can you share what logic you want your proxy to handle?

**Comment by haganbt** [comment]
Thanks @sidorares. Here is an example; we would like to pass through the auth from the client, through the proxy to the destination DB. 

```javascript
server.on('connection', async conn => {
  let remoteConn = null

  conn.serverHandshake({
    protocolVersion: 10,
    serverVersion: '5.7',
    connectionId: connectionId++,
    statusFlags: 2,
    characterSet: 8,
    capabilityFlags: 0xffffff ^ ClientFlags.COMPRESS,
    authCallback: async ({
      user,
      database,
      authToken,
      authPluginData1,
      authPluginData2,
    }) => {
      try {
        remoteConn = mysql.createConnection({
          database,
          host: 'arbitrary-host',
          port: 3306,
          user,
          password: '<USE PASSWORD FROM INCOMING CONNECTION>',
        })

        conn.writeOk()
      } catch (e) {
        conn.writeError(e)
      }
    },
  })

  conn.on('query', async sql => {
    remoteConn.query(sql, (e, results, fields) => {
      if (e) {
        conn.writeError(e)
      } else {
        conn.writeTextResult(results, fields)
      }
    })
  })
})
```

We dont need to do any validation or accessing of the password at the proxy, we simply wish to send it on to the destination database.

We would be very grateful if you can advise on a solution.
